### ISSUE TITLE
Extremely slow save callbacks go#lsp#Format and go#lsp#Imports

### ISSUE THREAD

**Issue by samsondav** [fatih/vim-go#3434]
### What did you do? (required: The issue will be **closed** when not provided)

Saved a go file with `:w`.

### What did you expect to happen?

I expected it to save quickly.

### What happened instead?

It saved very slowly and locked up editor until save was complete.

Here is the truncated profile:

```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    1  20.104363   0.000979  go#auto#fmt_autosave()
    2  20.103281   0.000261  go#fmt#Format()
    2  20.099677  20.099613  <SNR>150_await()
    1  10.052010   0.001172  go#lsp#Imports()
    1  10.050927   0.000283  go#lsp#Format()
    1   0.165254   0.000033  goimports#AutoRun()
    1   0.165221   0.000394  goimports#Run()
    1   0.164817   0.000908  <SNR>152_goimports()
```

#### vim-go version:

b7506c6da8ea3a88e268a91aa6630be3e26a20a9

####  `vimrc` you used to reproduce:

```vim
call plug#begin('~/.vim/plugged')
Plug 'fatih/vim-go', { 'do': ':GoUpdateBinaries' }
call plug#end()
```

#### Vim version (first three lines from `:version`):
```
NVIM v0.7.0
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by brew@HMBRW-A-001-M1-004.local
```
#### Go version (`go version`):
`go version go1.18.4 darwin/arm64`

#### Go environment
```GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/Users/sam/Library/Caches/go-build"
GOENV="/Users/sam/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/sam/.asdf/installs/golang/1.18.4/packages/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/sam/.asdf/installs/golang/1.18.4/packages"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/Users/sam/.asdf/installs/golang/1.18.4/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/Users/sam/.asdf/installs/golang/1.18.4/go/pkg/tool/darwin_arm64"
GOVCS=""
GOVERSION="go1.18.4"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/sam/code/smartcontractkit/chainlink/go.mod"
GOWORK="/Users/sam/code/smartcontractkit/go.work"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/z8/1hk864pd40lb5hfz5ljb75q00000gp/T/go-build2485438683=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

#### gopls version
```golang.org/x/tools/gopls v0.9.1
    golang.org/x/tools/gopls@v0.9.1 h1:SigsTL4Hpv3a6b/a7oPCLRv5QUeSM6PZNdta1oKY4B0=```

**Comment by bhcleek** [comment]
The time here is probably being spent in `gopls`. We may be able to get to the bottom of this.

1. Do you see this kind of poor performance even in very small modules (it's the module size, not the package size I'm wondering about)?
2. What is the output of the lsp log? You can turn on some lsp logging by setting `g:go_debug=['lsp']`.

**Comment by cippaciong** [comment]
I have the same issue and it happens even on a newly created module, with just a simple hello world.

I'm writing from the phone right now, I'll try to provide lsp logs once I'm back on the code.

**Comment by bhcleek** [comment]
There's a good chance that a 10 second delay on each of the operations that rely on `gopls` are due to `gopls` crashing. 

A work around until we can get to the bottom of your problem @samsondav is to use `goimports` for `g:go_fmt_command` and `g:go_imports_mode`. That may not be a long term solution, though, because using `goimports` comes with some tradeoffs. It's generally a great application, but it will be a bit slower due to how vim-go has to reload the file. Also, if `gopls` is crashing, then we'll want to see why so that it can be used for the other purposes that vim-go leverages it.

**Comment by cippaciong** [comment]
Figured out what was wrong with my setup. (TL;DR: I was using an ancient version of `gopls`)

First of all I wasn't able to reproduce the error with a fresh configuration of vim and the plugin. I then tried to enable debug logging as suggested by @bhcleek but I got an error from `gopls` not recognizing the `-debug` flag. Digging a little deeper I noticed that in my vim config I was using `Plug 'fatih/vim-go', { 'do': ':GoInstallBinaries' }` instead of `Plug 'fatih/vim-go', { 'do': ':GoUpdateBinaries' }`.

When I tried running`gopls version` it gave me an error but the creation date of the binary was `May 2019`. Considering that `gopls v0.1.0` is from `June 2019` I'm not even sure what I was running 😅 

Updating the binary with `GoUpdateBinaries` unsurprisingly fixed the issue.
